catroid/src/main/java/org/catrobat/catroid/io/asynctask/ProjectLoader.kt

```diff
@@ -0,0 +1,94 @@
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * An additional term exception under section 7 of the GNU Affero
+ * General Public License, version 3, is available at
+ * http://developer.catrobat.org/license_additional_term
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.catrobat.catroid.io.asynctask
+
+import android.content.Context
+import android.util.Log
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import org.catrobat.catroid.ProjectManager
+import org.catrobat.catroid.exceptions.ProjectException
+import org.catrobat.catroid.io.DeviceListAccessor
+import org.catrobat.catroid.io.DeviceVariableAccessor
+import org.catrobat.catroid.io.LookFileGarbageCollector
+import java.io.File
+import java.lang.ref.WeakReference
+
+class ProjectLoader {
+    private var projectDir: File? = null
+    private var weakContextReference: WeakReference<Context?>
+    private var weakListenerReference: WeakReference<ProjectLoadListener>? = null
+
+    companion object {
+        val TAG = ProjectLoader::class.java.simpleName
+    }
+
+    constructor(context: Context, listener: ProjectLoadListener) {
+        weakContextReference = WeakReference(context)
+        weakListenerReference = WeakReference(listener)
+    }
+
+    constructor(projectDir: File?, context: Context?) {
+        this.projectDir = projectDir
+        weakContextReference = WeakReference(context)
+    }
+
+    fun setListener(listener: ProjectLoadListener): ProjectLoader {
+        weakListenerReference = WeakReference(listener)
+        return this
+    }
+
+    @JvmOverloads
+    fun loadProjectAsync(scope: CoroutineScope = CoroutineScope(Dispatchers.IO)) {
+        val context = weakContextReference.get() ?: return
+        scope.launch {
+            val projectSaved = loadProject(projectDir, context)
+
+            withContext(Dispatchers.Main) {
+                val listener = weakListenerReference?.get() ?: return@withContext
+                listener.onLoadFinished(projectSaved)
+            }
+        }
+    }
+
+    interface ProjectLoadListener {
+        fun onLoadFinished(success: Boolean)
+    }
+}
+
+fun loadProject(projectDir: File?, context: Context?): Boolean {
+    return try {
+        ProjectManager.getInstance().loadProject(projectDir, context)
+        val project = ProjectManager.getInstance().currentProject
+        DeviceVariableAccessor(projectDir).cleanUpDeletedUserData(project)
```

Comment:
I commented out the calls to these three components to see if there are tests that check whether or not the clean up happens during the loading of a project. At least the in Unit Tests and Instrumented Unit Tests there weren't any failing tests.
I think this should be covered.